### PR TITLE
CI: caching tweaks in GitHub actions

### PR DESCRIPTION
**Pull request by kalessil**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

- Adds caching for Playwright
- Tweak composer caching condition

Existing CI is well optimized already: pnpm caching and build is already there (can be affected by GitHub CDN overload issues,  randomly causing full download/build activities).

### How to test the changes in this Pull Request:

- All CI checks shall pass

### CODE REVIEW

File: .github/actions/setup-woocommerce-monorepo/action.yml [hunk 1]
```diff
@@ -21,6 +21,35 @@ runs:
           uses: 'pnpm/action-setup@a3252b78c470c02df07e9d59298aecedc3ccdd6d'
           with:
               version: ${{ steps.read-pnpm-version.outputs.version }}
+        - name: 'Get PNPM store directory'
```
Comment (kalessil) [inline .github/actions/setup-woocommerce-monorepo/action.yml]: PNPM cache can be removed as CI already covers this:

```
/home/runner/setup-pnpm/node_modules/.bin/pnpm store path --silent
/home/runner/setup-pnpm/node_modules/.bin/store/v3
Received 188743680 of 712800998 (26.5%), 179.8 MBs/sec
Received 469762048 of 712800998 (65.9%), 223.9 MBs/sec
Cache Size: ~680 MB (712800998 B)
```

File: .github/actions/setup-woocommerce-monorepo/action.yml [hunk 2]
```diff
@@ -21,6 +21,35 @@ runs:
...
+          run: |
+              echo "PNPM_STORE_PATH=$(pnpm store path --silent)" >> $GITHUB_ENV
+        - name: 'Cache PNPM Dependencies'
+          if: ${{ inputs.build == 'false' }}
+          uses: 'actions/cache@ab5e6d0c87105b4c9c2047343972218f562e4319'
+          with:
+            path: ${{ env.PNPM_STORE_PATH }}
+            key: "${{ runner.os }}-pnpm-${{ hashFiles( '**/pnpm-lock.yaml' ) }}"
+            restore-keys: '${{ runner.os }}-pnpm-'
+        - name: 'Cache Playwright Downloads'
```
Comment (kalessil) [inline .github/actions/setup-woocommerce-monorepo/action.yml]: Playwright cache: to keep

File: .github/actions/setup-woocommerce-monorepo/action.yml [hunk 3]
```diff
@@ -21,6 +21,35 @@ runs:
...
+          if: ${{ inputs.build != 'false' }}
+          uses: 'actions/cache@ab5e6d0c87105b4c9c2047343972218f562e4319'
+          with:
+            path: '~/.cache/ms-playwright/'
+            key: "${{ runner.os }}-playwright-${{ hashFiles( '**/pnpm-lock.yaml' ) }}"
+            restore-keys: '${{ runner.os }}-playwright-'
+        - name: 'Cache build folders (mimics incremental builds)'
```
Comment (kalessil) [inline .github/actions/setup-woocommerce-monorepo/action.yml]: It is not needed; build times are still varying as they were before. There are no visible results for keeping this cache.

**Comment by rrennick** [comment]
It seems odd but the runtime on this PR for PHP unit tests is 12:45 which is longer than the [existing runtime](https://github.com/woocommerce/woocommerce/actions/runs/9688522998/job/26735282726?pr=48878) on other PRs due to a few steps taking longer.

File: .github/actions/setup-woocommerce-monorepo/action.yml [hunk 4]
```diff
@@ -34,7 +34,7 @@ runs:
               php-version: '${{ inputs.php-version }}'
               coverage: 'none'
         - name: 'Cache Composer Dependencies'
-          if: ${{ inputs.php-version != 'false' }}
+          if: ${{ inputs.build == 'false' }}
```
Comment (kalessil) [inline .github/actions/setup-woocommerce-monorepo/action.yml:37]: Ensures the cache is always available: for pnpm install / building the project.

**Comment by kalessil** [comment]
Update:

> Existing CI is well optimized already: pnpm caching and build is already there 
> (can be affected by GitHub CDN overload issues, randomly causing full download/build activities).

I'm leaving some tweaks, which are minor improvements (positive impact is magnitude of seconds).

**Comment by kalessil** [comment]
We are good for code & build times review here. 

I cleaned up earlier experiments (as @rrennick mentioned, build times only increased), but I found two minor improvements:
- Playwright downloads caching
- Enabling pnpm deps caching in building live branches (hope the way I did it is ok)

So reviewers are welcome to poke the changes with a stick and request reverting changes if those don't woth keeping.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -135,6 +135,15 @@ jobs:
...
               run: node ./tests/e2e-pw/bin/override-wp-env-plugins.js
+
+            - name: 'Cache Playwright Downloads'
+              id: 'cache-playwright-downloads'
+              if: ${{ matrix.testEnv.shouldCreate }}
+              uses: 'actions/cache@ab5e6d0c87105b4c9c2047343972218f562e4319'
+              with:
+                path: '~/.cache/ms-playwright/'
+                key: "${{ runner.os }}-playwright-${{ hashFiles( '**/pnpm-lock.yaml' ) }}"
+                restore-keys: '${{ runner.os }}-playwright-'
```
Comment (lanej0) [inline .github/workflows/ci.yml:146]: Solid improvement here.  It looks like it takes it from >1 min to 20 sec.

File: .github/workflows/build-live-branch.yml [hunk 6]
```diff
@@ -31,6 +31,12 @@ jobs:
             - name: Setup WooCommerce Monorepo
               uses: ./.github/actions/setup-woocommerce-monorepo
 
+            - name: Cache PNPM Dependencies
+              uses: actions/setup-node@8f152de45cc393bb48ce5d89d36b731f54556e65
+              with:
+                node-version-file: .nvmrc
+                cache: pnpm
+
```
Comment (lanej0) [inline .github/workflows/build-live-branch.yml:39]: Again, solid improvement here.  Down to 6 minutes from 13 or 14 (it looks like).

**Review by lanej0 — APPROVED** [review]
Solid improvements to timings through caching here.  Looks awesome!

File: .github/workflows/ci.yml (hunk 5 above)
Comment (adimoldovan) [reply .github/workflows/ci.yml:146]: What is this caching exactly? Is it the browsers binaries?

Comment (adimoldovan) [reply .github/workflows/ci.yml:146]: If it's the browser's binaries shouldn't the caching happen after the env setup step? That's where `playwright install` runs.

Comment (kalessil) [reply .github/workflows/ci.yml:146]: Yes, it's chromium plus ffmpeg binaries. 

It's placed before as we try to restore the cache before Playwright downloading anything.
A cache update will be attempted in `Post Cache Playwright Downloads` step (automatic post run hook).

File: .github/workflows/build-live-branch.yml (hunk 6 above)
Comment (kalessil) [reply .github/workflows/build-live-branch.yml:39]: Oh, I wish it was that good =) I observed around 10 seconds of improvements (excluding cache restore time), but we should get more consistent with build times (the time needed to download all packages might differ).

IIRC we have some stats related to builds, hope we can pull some numbers next week and see the impact of the changes.

**Pull request merged**